File: EmarsysPredictShop/EmarsysPredictShop/LoginViewController.m
//
//  LoginViewController.m
//  EmarsysPredictShop
//

#import <EmarsysPredictSDK/EmarsysPredictSDK.h>
@import AdSupport.ASIdentifierManager;

#import "LoginViewController.h"

@interface LoginViewController ()

@end

NSString *const LoginCustomerEmail = @"customer_email";
NSString *const LoginCustomerId = @"customer_id";
NSString *const ItemDetailLogic = @"item_detail_logic";

@implementation LoginViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.isLogin = false;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];

    NSString *email = [defaults objectForKey:LoginCustomerEmail];
    if (email) {
        [self.emailTextField setText:email];
    }
    NSString *id = [defaults objectForKey:LoginCustomerId];
    if (id) {
        [self.customerIdTextField setText:id];
    }

    NSString *itemLogic = [defaults objectForKey:ItemDetailLogic];
    if (itemLogic) {
        NSInteger i = 0;
        if ([itemLogic isEqualToString:@"RELATED"]) {
            i = 0;
        } else if ([itemLogic isEqualToString:@"ALSO_BOUGHT"]) {
            i = 1;
        }
        [self.itemRecommendedLogic setSelectedSegmentIndex:i];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)login:(id)sender {

    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    BOOL loginWithEmail = _enableLoginWithEmail.isOn;
    EMSession *emarsysSession = [EMSession sharedSession];
    NSString *message = nil;

    if (self.isLogin) {
        [self.loginButton setTitle:@"Login"];
        self.isLogin = false;
        [emarsysSession setCustomerEmail:nil];
        [emarsysSession setCustomerID:nil];
        message = @"Logout successful";
    } else {
        if (loginWithEmail) {
            NSString *email = [defaults objectForKey:LoginCustomerEmail];
            if (email) {
                [emarsysSession setCustomerEmail:email];
                message = @"Login was successful";
                [self.loginButton setTitle:@"Logout"];
                self.isLogin = true;
            } else {
                message = @"Please set email";
            }
        } else {
            NSString *customerId = [defaults objectForKey:LoginCustomerId];
            if (customerId) {
                [emarsysSession setCustomerID:customerId];
                message = @"Login was successful";
                [self.loginButton setTitle:@"Logout"];
                self.isLogin = true;
            } else {
                message = @"Please set customer id";
            }
        }
    }

    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:message
                         message:nil
                  preferredStyle:UIAlertControllerStyleAlert];

    UIAlertAction *defaultAction =
        [UIAlertAction actionWithTitle:@"OK"
                                 style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction *action){
                               }];
    [alert addAction:defaultAction];
    [self presentViewController:alert animated:YES completion:nil];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField;
{
    [textField resignFirstResponder];
    return YES;
}

- (IBAction)save:(id)sender {

    // Store the data
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    if (self.emailTextField.hasText) {
        /// Set customer email address.
        [defaults setObject:self.emailTextField.text forKey:LoginCustomerEmail];
    }

    if (self.customerIdTextField.hasText) {
        /// Set customer id.
        [defaults setObject:self.customerIdTextField.text
                     forKey:LoginCustomerId];
    }

    NSInteger selectedSegment = _itemRecommendedLogic.selectedSegmentIndex;

    NSString *logic;

    if (selectedSegment == 0) {
        logic = @"RELATED";
    } else {
        logic = @"ALSO_BOUGHT";
    }

    [defaults setObject:logic forKey:ItemDetailLogic];
    [defaults synchronize];

    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:@"Save was successful"
                         message:nil
                  preferredStyle:UIAlertControllerStyleAlert];

    UIAlertAction *defaultAction =
        [UIAlertAction actionWithTitle:@"OK"
                                 style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction *action){
                               }];
    [alert addAction:defaultAction];
    [self presentViewController:alert animated:YES completion:nil];
}

- (IBAction)showIDs:(id)sender {
    NSString *message =
        [NSString stringWithFormat:@"IDFA:\n%@\n\nAdvertising UUID:\n%@",
                                   [[ASIdentifierManager sharedManager]
                                       advertisingIdentifier]
                                       .UUIDString,
                                   [[EMSession sharedSession] advertisingID]];
    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:message
                         message:nil
                  preferredStyle:UIAlertControllerStyleAlert];

    UIAlertAction *defaultAction =
        [UIAlertAction actionWithTitle:@"OK"
                                 style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction *action){
                               }];
    [alert addAction:defaultAction];
    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: EmarsysMobileShopSwift/EmarsysMobileShopSwift/LoginViewController.swift
//
//  LoginViewController.swift
//  EmarsysMobileShopSwift
//
//

import UIKit
import EmarsysMobile
import AdSupport

enum UserDefaultsKeys : String {
    case Email = "customer_email"
    case ID = "customer_id"
    case DetailLogic = "item_detail_logic"
}

class LoginViewController: UIViewController, UITextFieldDelegate {
    
    @IBOutlet weak var emailTextField: UITextField!
    @IBOutlet weak var customerIdTextField: UITextField!
    @IBOutlet weak var itemRecommendedLogic: UISegmentedControl!
    @IBOutlet weak var saveButton: UIButton!
    @IBOutlet weak var enableLoginWithEmail: UISwitch!
    @IBOutlet weak var loginButton: UIBarButtonItem!
    
    var isLogin: Bool = false
    
    @IBAction func login(sender: UIBarButtonItem) {
        
        let defaults = NSUserDefaults.standardUserDefaults()
        let emarsysSession = EMSession.sharedSession()
        let loginWithEmail = enableLoginWithEmail.on
        var message: String!
        
        if isLogin {
            loginButton.title = "Login"
            isLogin = false
            emarsysSession.customerEmail = nil
            emarsysSession.customerID = nil
            message = "Logout successful"
        } else if loginWithEmail {
            if let email = defaults.objectForKey(UserDefaultsKeys.Email.rawValue) {
                emarsysSession.customerEmail = email as? String
                message = "Login was successful"
                loginButton.title = "Logout"
                isLogin = true
            } else {
                message = "Please set email"
            }
        } else {
            if let customerId = defaults.objectForKey(UserDefaultsKeys.ID.rawValue) {
                emarsysSession.customerID = customerId as? String
                message = "Login was successful"
                loginButton.title = "Logout"
                isLogin = true
            } else {
                message = "Please set customer id"
            }
        }
        
        let alert = UIAlertController(title: message, message: nil, preferredStyle: .Alert)
        let defaultAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: {(action) in })
        alert.addAction(defaultAction)
        presentViewController(alert, animated: true, completion: nil)
    }
    
    @IBAction func save(sender: UIButton) {
        
        // Store the data
        let defaults = NSUserDefaults.standardUserDefaults()
        if let email = emailTextField.text {
            /// Set customer email address.
            defaults.setObject(email, forKey: UserDefaultsKeys.Email.rawValue)
        }
        
        if let customerId = customerIdTextField.text {
            /// Set customer id.
            defaults.setObject(customerId, forKey: UserDefaultsKeys.ID.rawValue)
        }
        
        let selectedSegment = itemRecommendedLogic.selectedSegmentIndex
        let logic = selectedSegment == 0 ? "RELATED" : "ALSO_BOUGHT"
        
        defaults.setObject(logic, forKey: UserDefaultsKeys.DetailLogic.rawValue)
        defaults.synchronize()
        
        let alert = UIAlertController(title: "Save was successful", message: nil, preferredStyle: .Alert)
        let defaultAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: {(action) in })
        alert.addAction(defaultAction)
        presentViewController(alert, animated: true, completion: nil)
    }
    
    @IBAction func showIDs(sender: AnyObject) {
        let alert = UIAlertController(title: "IDFA:\n\(ASIdentifierManager.sharedManager().advertisingIdentifier?.UUIDString ?? "nil")\n\nAdvertising UUID:\n\(EMSession.sharedSession().advertisingID)", message: nil, preferredStyle: .Alert)
        let defaultAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: {(action) in })
        alert.addAction(defaultAction)
        presentViewController(alert, animated: true, completion: nil)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        let defaults = NSUserDefaults.standardUserDefaults()
        
        if let email = defaults.objectForKey(UserDefaultsKeys.Email.rawValue) {
            emailTextField.text = email as? String
        }
        
        if let id = defaults.objectForKey(UserDefaultsKeys.ID.rawValue) {
            customerIdTextField.text = id as? String
        }
        
        if let userLogic = defaults.objectForKey(UserDefaultsKeys.DetailLogic.rawValue) {
            let logic = userLogic as! String
            itemRecommendedLogic.selectedSegmentIndex = (logic == "ALSO_BOUGHT" ? 1 : 0)
        }
    }
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
}
